### PR TITLE
Add mock SFT dataset and varlen dataset family

### PR DESCRIPTION
**Pull request by ilml**
## What does this PR do?

Adds the mock SFT dataset (with a generalized SFT padding divisor) and the varlen dataset family (`VarlenLowLevelDataset`, `VarlenDataset`, `MockVarlenDataset`) with unit tests. The datasets are inert until the follow-up training-integration PR wires them into pretraining.

Part of the [#3386](https://github.com/NVIDIA/Megatron-LM/pull/3386) split series tracked in the [Dynamic Context Parallelism project](https://github.com/orgs/NVIDIA/projects/297). Original changes by @xiaoyao0115 in #3386.

**Series and merge order** (reviews can run in parallel):

| PR | Content | Merge order |
|---|---|---|
| #5902 | config field, CLI arg, validation | 1 (independent) |
| #6684 | core packing scheduler | 1 (independent) |
| **this PR** | mock SFT + varlen datasets | 2 — after #6684 (stacked on it) and #5902 |
| training-integration PR | scheduler + varlen wiring into training loop | 3 — last |

> **Stacking notes:**
> - Base is `pull-request/6684` so Files-changed shows only this PR's layers. It will be retargeted to `main` as soon as #6684 is approved, before any merge.
> - The first commit is a temporary copy of #5902's change (the arguments.py context depends on it); it drops out automatically when #5902 merges and this branch is rebased.

## Contribution process

- [x] Draft PR per contributing guidelines
- [x] Commits signed off (DCO)

### CODE REVIEW

**Review by zhongbozhu — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -3200,6 +3200,40 @@ def _scope_to_str(s):
                     "Disable MoE capacity/expert padding."
                 )
 
+        if self.sequence_packing_scheduler is not None:
```
Comment (jaredcasper) [inline megatron/core/transformer/transformer_config.py:3203]: Shouldn't some/most of these checks be in model_parallel_config if that's where sequence_packing_scheduler is?

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -3719,8 +3753,28 @@ def _add_kitchen_quantization_arguments(parser: argparse.ArgumentParser):
 def _add_sft_args(parser):
     group = parser.add_argument_group(title='sft')
     group.add_argument('--sft', action="store_true", help='Megatron SFT training')
-    group.add_argument('--sft-tokenizer-prompt-format', type=str, default="nemotron-h-aligned",
-                       help='SFT prompt format.')
+    group.add_argument(
```
Comment (jaredcasper) [inline megatron/training/arguments.py:3756]: Why change the formatting here?

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -3719,8 +3753,28 @@ def _add_kitchen_quantization_arguments(parser: argparse.ArgumentParser):
...
+        '--sft-tokenizer-prompt-format',
+        type=str,
+        default="nemotron-h-aligned",
+        help='SFT prompt format.',
+    )
+    group.add_argument(
```
Comment (jaredcasper) [inline megatron/training/arguments.py:3762]: Can you match the formatting of other arguments?

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -1642,6 +1652,19 @@ def validate_args(args, defaults={}):
     if args.ckpt_format == "fsdp_dtensor":
         assert args.use_megatron_fsdp, "--ckpt-format fsdp_dtensor is only tested with Megatron FSDP."
 
+    # Packed-sequence buffer-size check. Placed after varlen scheduler
+    # auto-select so it validates the final resolved scheduler.
+    if args.sequence_packing_scheduler is not None:
```
Comment (jaredcasper) [inline megatron/training/arguments.py:1657]: Can these checks be in the config dataclass post_init?
